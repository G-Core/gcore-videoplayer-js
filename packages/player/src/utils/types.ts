import { $ } from "@clappr/core";

/**
 * {@link https://zeptojs.com/#$() | Zepto query result}
 * @beta
 */
export type ZeptoResult = ReturnType<typeof $>;

/**
 * @internal
 */
export type TimerId = ReturnType<typeof setTimeout>;
