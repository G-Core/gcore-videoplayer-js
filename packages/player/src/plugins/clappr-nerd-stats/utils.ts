import { ZeptoResult } from "../../types"
import { SpeedtestMetrics } from "./speedtest/types"

const qualityClasses = [
  'speedtest-quality-value-1',
  'speedtest-quality-value-2',
  'speedtest-quality-value-3',
  'speedtest-quality-value-4',
  'speedtest-quality-value-5',
]

export const getDownloadQuality = (speedValue: number): number => {
  if (speedValue < 3) {
    return 1
  } else if (speedValue < 7) {
    return 2
  } else if (speedValue < 13) {
    return 3
  } else if (speedValue < 25) {
    return 4
  } else {
    return 5
  }
}

export const getPingQuality = (pingValue: number): number => {
  if (pingValue < 20) {
    return 5
  } else if (pingValue < 50) {
    return 4
  } else if (pingValue < 100) {
    return 3
  } else if (pingValue < 150) {
    return 2
  } else {
    return 1
  }
}

export const generateQualityHtml = (quality: number): string => {
  const html = []
  const qualityClassName = qualityClasses[quality - 1]

  for (let i = 0; i < qualityClasses.length; i++) {
    if (i < quality) {
      html.push(
        `<div class="speedtest-quality-content-item ${qualityClassName}"></div>`,
      )
    } else {
      html.push('<div class="speedtest-quality-content-item"></div>')
    }
  }

  return html.join('')
}

export const drawSummary = (
  customMetrics: SpeedtestMetrics,
  vodContainer: ZeptoResult,
  liveContainer: ZeptoResult,
) => {
  const { connectionSpeed, ping } = customMetrics

  if (!connectionSpeed || !ping) {
    return
  }
  const downloadQuality = getDownloadQuality(connectionSpeed)
  const pingQuality = getPingQuality(ping)
  const liveQuality = Math.min(downloadQuality, pingQuality)
  const vodHtml = generateQualityHtml(downloadQuality)
  const liveHtml = generateQualityHtml(liveQuality)

  vodContainer.html(vodHtml)
  liveContainer.html(liveHtml)
}
