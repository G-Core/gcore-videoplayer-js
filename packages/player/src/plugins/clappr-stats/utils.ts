import type { Metrics } from './types'

export function newMetrics(): Metrics {
  return {
    counters: {
      play: 0,
      pause: 0,
      error: 0,
      buffering: 0,
      decodedFrames: 0,
      droppedFrames: 0,
      fps: 0,
      changeLevel: 0,
      seek: 0,
      fullscreen: 0,
      dvrUsage: 0,
    },
    chrono: {
      startup: 0,
      watch: 0,
      pause: 0,
      buffering: 0,
      session: 0,
    },
    extra: {
      playbackName: '',
      playbackType: '',
      bitratesHistory: [],
      bitrateWeightedMean: 0,
      bitrateMostUsed: 0,
      buffersize: 0,
      watchHistory: [],
      watchedPercentage: 0,
      bufferingPercentage: 0,
      bandwidth: 0,
      duration: 0,
      currentTime: 0,
    },
  }
}
