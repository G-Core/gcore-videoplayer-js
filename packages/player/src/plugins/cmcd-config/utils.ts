export function generateSessionId(): string {
  return window.crypto.randomUUID()
}

export function generateContentId(sourceUrl: string): Promise<string> {
  return window.crypto.subtle.digest('SHA-1', new TextEncoder().encode(sourceUrl))
    .then(buffer => {
      const hex = Array.from(new Uint8Array(buffer))
        .map(b => b.toString(16).padStart(2, '0'))
        .join('')
      return hex
    })
}
