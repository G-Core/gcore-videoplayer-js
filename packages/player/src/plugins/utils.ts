export function getLocation(href: string) {
  const l = document.createElement('a');

  l.href = href;

  return l;
}

export function strtimeToMiliseconds(str: string): number {
  if (!str) {
    return 0;
  }
  const arr = str.split(/:/);
  let h = 0, m = 0, s = 0;

  if (arr.length >= 3) {
    h = parseInt(arr[arr.length - 3]) * 60 * 60;
  } else {
    h = 0;
  }
  if (arr.length >= 2) {
    m = parseInt(arr[arr.length - 2]) * 60;
  } else {
    m = 0;
  }

  if (arr.length >= 1) {
    s = parseInt(arr[arr.length - 1]);
  } else {
    s = 0;
  }

  return (h + m + s);
}

export function getPageX(event: MouseEvent | TouchEvent): number {
  if ((event as MouseEvent).pageX) {
    return (event as MouseEvent).pageX;
  }

  if ((event as TouchEvent).changedTouches) {
    return (event as TouchEvent).changedTouches[(event as TouchEvent).changedTouches.length - 1].pageX;
  }

  return 0;
}
