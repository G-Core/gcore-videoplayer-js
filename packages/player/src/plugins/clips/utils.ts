import { ClipDesc } from './types.js'
import { parseClipTime } from '../utils.js'

type ClipItem = {
  start: number
  text: string
}

export function parseClips(text: string): ClipDesc[] {
  const clipsArr = text
    .split('\n')
    .map((val: string) => {
      const matchRes = val.match(/(((\d+:)?\d+:)?\d+) (.+)/i)
      return matchRes
        ? {
            start: parseClipTime(matchRes[1]),
            text: matchRes[4],
          }
        : null
    })
    .filter((clip: ClipItem | null) => clip !== null)
    .sort((a: ClipItem, b: ClipItem) => a.start - b.start)
  return clipsArr.map((clip: ClipItem, index: number) => ({
    start: clip.start,
    text: clip.text,
    end: index < clipsArr.length - 1 ? clipsArr[index + 1].start : 0,
  }))
}

export function buildSvg(clips: ClipDesc[], duration: number, barWidth: number): string {
  let svg =
    '<svg width="0" height="0">\n' + '<defs>\n' + '<clipPath id="myClip">\n'
  let rightEdge = 0

  clips.forEach((val) => {
    const end = val.end || duration

    const chunkWidth = Math.round(((end - val.start) * barWidth) / duration)

    svg += `<rect x="${rightEdge}" y="0" width="${
      chunkWidth - 2
    }" height="30"/>\n`
    rightEdge += chunkWidth
  })

  if (rightEdge < barWidth) {
    svg += `<rect x="${rightEdge}" y="0" width="${
      barWidth - rightEdge
    }" height="30"/>\n`
  }
  svg += '</clipPath>' + '</defs>' + '</svg>'

  return svg
}
