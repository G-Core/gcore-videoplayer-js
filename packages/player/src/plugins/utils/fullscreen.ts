import { Browser, Utils } from '@clappr/core';
import { reportError } from '@gcorevideo/utils';

export const fullscreenEnabled = Utils.Fullscreen.fullscreenEnabled

export function isFullscreen(el: HTMLElement): boolean {
  const video =
    el.nodeName === 'video'
      ? (el as HTMLVideoElement)
      : el.querySelector('video')
  if (!video) {
    return false
  }
  if (Browser.isiOS) {
    return FullscreenIOS.isFullscreen(video)
  }
  return !!document.fullscreenElement
}

const FullscreenIOS = {
  isFullscreen: function (el: HTMLVideoElement): boolean {
    try {
      // @ts-ignore
      if (el.webkitDisplayingFullscreen !== undefined) {
        // @ts-ignore
        return !!el.webkitDisplayingFullscreen
      }
    } catch (e) {
      reportError(e)
    }

    return false
  },
}
