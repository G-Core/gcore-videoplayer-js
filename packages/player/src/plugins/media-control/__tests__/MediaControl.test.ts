import { describe, it, expect, vi, beforeEach } from 'vitest'
import {
  MediaControl,
  MediaControlElement,
  MediaControlSettings,
} from '../MediaControl'
import { createMockCore } from '../../../testUtils'
import { LogTracer, Logger, setTracer } from '@gcorevideo/utils'
import { $, Events, Playback } from '@clappr/core'

vi.mock('../../utils/fullscreen', () => ({
  fullscreenEnabled: vi.fn().mockReturnValue(true),
  isFullscreen: vi.fn().mockReturnValue(false),
}))

Logger.enable('*')
setTracer(new LogTracer('MediaControl.test'))

describe('MediaControl', () => {
  let core: any
  let mediaControl: MediaControl

  beforeEach(() => {
    core = createMockCore()
  })
  describe('initially', () => {
    beforeEach(() => {
      mediaControl = new MediaControl(core)
      core.emit(Events.CORE_READY)
    })
    it('should not render', () => {
      expect(mediaControl.el.innerHTML).toBe('')
    })
  })
  describe('rendering timing', () => {
    beforeEach(() => {
      mediaControl = new MediaControl(core)
      core.emit(Events.CORE_READY)
      core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
      core.activeContainer.settings = {
        left: ['playpause', 'position', 'duration', 'volume'],
        default: ['seekbar'],
        right: ['fullscreen', 'hd-indicator'],
        seekEnabled: true,
      } as MediaControlSettings
      core.activePlayback.emit(Events.PLAYBACK_SETTINGSUPDATE)
      core.activeContainer.emit(Events.CONTAINER_SETTINGSUPDATE)
    })
    describe('until metadata is loaded', () => {
      it('should not render', () => {
        expect(mediaControl.el.innerHTML).toBe('')
      })
    })
    describe('once metadata is loaded', () => {
      beforeEach(() => {
        core.activePlayback.emit(Events.PLAYBACK_LOADEDMETADATA)
        core.activeContainer.emit(Events.CONTAINER_LOADEDMETADATA)
      })
      it('should wait a delay before rendering anything', async () => {
        expect(mediaControl.el.innerHTML).toBe('')
        await new Promise((resolve) => setTimeout(resolve, 35))
        expect(mediaControl.el.innerHTML).toMatchSnapshot()
        expect(
          mediaControl.$el.find('.media-control-left-panel [data-playpause]')
            .length,
        ).toBeGreaterThan(0)
        expect(
          mediaControl.$el.find('.media-control-left-panel [data-volume]')
            .length,
        ).toBeGreaterThan(0)
        expect(
          mediaControl.$el.find('.media-control-center-panel [data-seekbar]')
            .length,
        ).toBeGreaterThan(0)
        expect(
          mediaControl.$el.find('.media-control-right-panel [data-fullscreen]')
            .length,
        ).toBeGreaterThan(0)
      })
    })
  })
  describe('when container settings update', () => {
    beforeEach(async () => {
      mediaControl = new MediaControl(core)
      core.emit(Events.CORE_READY)
      core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
      await runMetadataLoaded(core)
    })
    describe.each([
      [
        'vod',
        {
          left: ['playpause', 'position', 'duration', 'volume'],
          default: ['seekbar'],
          right: ['fullscreen', 'hd-indicator'],
          seekEnabled: true,
        } as MediaControlSettings,
      ],
    ])('%s', (_, settings: MediaControlSettings) => {
      beforeEach(() => {
        core.activeContainer.settings = settings
        core.activePlayback.emit(Events.PLAYBACK_SETTINGSUPDATE)
        core.activeContainer.emit(Events.CONTAINER_SETTINGSUPDATE)
      })
      it.each(settings.left)('should render %s control', (element) => {
        const el = mediaControl.$el.find(
          `.media-control-left-panel [data-${element}]`,
        )
        expect(el.length).toBeGreaterThan(0)
      })
      it.each(
        arraySubtract(
          ['playpause', 'playstop', 'position', 'duration', 'volume'],
          settings.left,
        ),
      )('should not render %s control', (element) => {
        const el = mediaControl.$el.find(
          `.media-control-left-panel [data-${element}]`,
        )
        expect(el.length).toEqual(0)
      })
      it(`should ${
        settings.seekEnabled ? '' : 'not '
      }render the seek bar`, () => {
        const seekbar = mediaControl.$el.find(
          '.media-control-center-panel [data-seekbar]',
        )
        expect(seekbar.length).toBeGreaterThan(1)
        if (settings.seekEnabled) {
          expect(seekbar.hasClass('seek-disabled')).toBe(false)
        } else {
          expect(seekbar.hasClass('seek-disabled')).toBe(true)
        }
      })
      it(`should ${
        settings.left.includes('volume') ? '' : 'not '
      }render the volume control`, () => {
        const volume = mediaControl.$el.find('.drawer-container[data-volume]')
        if (settings.left.includes('volume')) {
          expect(volume.length).toEqual(1)
        } else {
          expect(volume.length).toEqual(0)
        }
      })
      it(`should ${
        settings.right.includes('fullscreen') ? '' : 'not '
      }render the fullscreen control`, () => {
        const fullscreen = mediaControl.$el.find(
          '.media-control-right-panel [data-fullscreen]',
        )
        if (settings.right.includes('fullscreen')) {
          expect(fullscreen.length).toEqual(1)
        } else {
          expect(fullscreen.length).toEqual(0)
        }
      })
    })
  })
  describe('playback type', () => {
    beforeEach(() => {
      mediaControl = new MediaControl(core)
      core.emit(Events.CORE_READY)
      core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
    })
    describe('when live', () => {
      beforeEach(async () => {
        core.activeContainer.getPlaybackType.mockReturnValue(Playback.LIVE)
        core.activePlayback.getPlaybackType.mockReturnValue(Playback.LIVE)
        core.getPlaybackType.mockReturnValue(Playback.LIVE)
        // This is not strictly necessary as the CSS class is applied on the root element and does not require rendering.
        // However, it makes the scenario more realistic
        await runMetadataLoaded(core)
        // TODO playback.settings
        core.activePlayback.emit(Events.PLAYBACK_SETTINGSUPDATE)
      })
      it('should apply live style class', () => {
        expect(mediaControl.$el.hasClass('live')).toBe(true)
      })
    })
    describe('when vod', () => {
      beforeEach(async () => {
        core.activeContainer.getPlaybackType.mockReturnValue(Playback.VOD)
        core.activePlayback.getPlaybackType.mockReturnValue(Playback.VOD)
        core.getPlaybackType.mockReturnValue(Playback.VOD)
        await runMetadataLoaded(core)
      })
      it('should not apply live style class', () => {
        expect(mediaControl.$el.hasClass('live')).toBe(false)
      })
    })
  })
  describe('mount', () => {
    beforeEach(async () => {
      mediaControl = new MediaControl(core)
      core.emit(Events.CORE_READY)
      core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
      core.activeContainer.settings = {}
      core.emit(Events.CONTAINER_SETTINGSUPDATE)
      await runMetadataLoaded(core)
    })
    describe.each([
      ['pip' as MediaControlElement],
      ['gear' as MediaControlElement],
      ['cc' as MediaControlElement],
      // ['multicamera' as MediaControlElement],
      // ['playbackrate' as MediaControlElement],
      // ['vr' as MediaControlElement],
      ['audiotracks' as MediaControlElement],
      // dvr controls
    ])('%s', (mcName) => {
      it('should put the element in the right panel', () => {
        const element = document.createElement('div')
        element.className = 'my-media-control'
        element.textContent = 'test'
        mediaControl.mount(mcName, $(element))

        expect(mediaControl.el.innerHTML).toMatchSnapshot()
        expect(
          mediaControl.$el.find('.media-control-right-panel .my-media-control')
            .length,
        ).toEqual(1)
      })
    })
  })
  describe('updateSettings', () => {
    beforeEach(() => {
      mediaControl = new MediaControl(core)
      core.emit(Events.CORE_READY)
    })
    describe('dvr', () => {
      beforeEach(async () => {
        core.activeContainer.settings = {
          left: ['playpause', 'position', 'duration'],
          seekEnabled: true,
        }
        core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
        core.activePlayback.getPlaybackType.mockReturnValue(Playback.LIVE)
        core.activeContainer.getPlaybackType.mockReturnValue(Playback.LIVE)
        core.getPlaybackType.mockReturnValue(Playback.LIVE)
        await runMetadataLoaded(core)
      })
      describe('when enabled', () => {
        beforeEach(() => {
          core.activePlayback.dvrEnabled = true
          core.activeContainer.isDvrEnabled.mockReturnValue(true)
          core.activeContainer.emit(Events.CONTAINER_SETTINGSUPDATE)
        })
        it('should enable DVR controls', () => {
          const element = document.createElement('div')
          element.className = 'my-dvr-controls'
          element.textContent = 'live'
          mediaControl.mount('dvr', $(element))
          expect(mediaControl.el.innerHTML).toMatchSnapshot()
          expect(
            mediaControl.$el.find('.media-control-left-panel .my-dvr-controls')
              .length,
          ).toEqual(1)
        })
      })
      describe('when disabled', () => {
        it('should disable DVR controls', () => {
          const element = document.createElement('div')
          element.className = 'my-dvr-controls'
          element.textContent = 'live'
          mediaControl.mount('dvr', $(element))
          expect(mediaControl.el.innerHTML).toMatchSnapshot()
          expect(
            mediaControl.$el.find('.media-control-left-panel .my-dvr-controls')
              .length,
          ).toEqual(0)
        })
      })
    })
  })
  describe('dvr mode', () => {
    beforeEach(async () => {
      mediaControl = new MediaControl(core)
      core.emit(Events.CORE_READY)
      core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
      await runMetadataLoaded(core)
    })
    describe('by default', () => {
      it('should not apply DVR style class', () => {
        expect(mediaControl.$el.hasClass('dvr')).toBe(false)
      })
    })
    describe('when in use', () => {
      beforeEach(() => {
        core.activeContainer.isDvrInUse.mockReturnValue(true)
        core.activePlayback.dvrInUse = true
        core.activeContainer.emit(
          Events.CONTAINER_PLAYBACKDVRSTATECHANGED,
          true,
        )
      })
      it('should apply DVR style class', () => {
        expect(mediaControl.$el.hasClass('dvr')).toBe(true)
      })
    })
  })
  describe('seekbar', () => {
    beforeEach(async () => {
      mediaControl = new MediaControl(core)
      core.emit(Events.CORE_READY)
      core.activeContainer.settings = {
        seekEnabled: true,
        default: ['seekbar'],
      }
      core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
      await runMetadataLoaded(core)
      core.activeContainer.emit(Events.CONTAINER_SETTINGSUPDATE)
    })
    it('should render', () => {
      expect(mediaControl.el.innerHTML).toMatchSnapshot()
    })
  })
})

function arraySubtract<T extends string>(arr1: T[], arr2: T[]) {
  return arr1.filter((item) => !arr2.includes(item))
}

function runMetadataLoaded(core: any) {
  core.activePlayback.emit(Events.PLAYBACK_LOADEDMETADATA)
  core.activeContainer.emit(Events.CONTAINER_LOADEDMETADATA)
  return new Promise((resolve) => setTimeout(resolve, 25))
}
