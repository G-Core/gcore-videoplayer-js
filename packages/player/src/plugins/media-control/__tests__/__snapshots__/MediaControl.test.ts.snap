// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MediaControl > putElement > audiotracks > should put the element in the right panel 1`] = `
"<div class="media-control-background" data-background=""></div>

<div class="media-control-layer gcore-skin-bg-color" data-controls="">
    
    
    
    
    
    
    
    
    

    
    
        <div class="media-control-left-panel" data-media-control="">
            
        <div class="drawer-container" data-volume="">
            <div class="drawer-icon-container" data-volume="">
                <div class="drawer-icon media-control-icon gcore-skin-button-color" data-volume="">/assets/icons/new/volume-max.svg</div>
                <span class="drawer-text" data-volume=""></span>
            </div>
            
    <div class="bar-container" data-volume="">
        <div class="bar-background" data-volume="">
            <div class="bar-fill-1 gcore-skin-main-color" data-volume="" style=""></div>
        </div>
        <div class="bar-scrubber" data-volume="" style="">
            <div class="bar-scrubber-icon gcore-skin-main-color" data-volume=""></div>
        </div>
    </div>
    
        </div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-clips=""></div>
    
        </div>
    
    
    
        <div class="media-control-right-panel" data-media-control="">
            
        <button type="button" class="media-control-button media-control-icon gcore-skin-button-color" data-fullscreen="" aria-label="fullscreen">/assets/icons/new/fullscreen-off.svg</button>
    
        <div class="media-control-multicamera" data-multicamera=""></div>
    
        <div class="media-control-vr" data-vr=""></div>
    
        <div class="my-media-control" data-audiotracks="">test</div></div>
    
</div>
<style>:root {}</style>"
`;

exports[`MediaControl > putElement > cc > should put the element in the right panel 1`] = `
"<div class="media-control-background" data-background=""></div>

<div class="media-control-layer gcore-skin-bg-color" data-controls="">
    
    
    
    
    
    
    
    
    

    
    
        <div class="media-control-left-panel" data-media-control="">
            
        <div class="drawer-container" data-volume="">
            <div class="drawer-icon-container" data-volume="">
                <div class="drawer-icon media-control-icon gcore-skin-button-color" data-volume="">/assets/icons/new/volume-max.svg</div>
                <span class="drawer-text" data-volume=""></span>
            </div>
            
    <div class="bar-container" data-volume="">
        <div class="bar-background" data-volume="">
            <div class="bar-fill-1 gcore-skin-main-color" data-volume="" style=""></div>
        </div>
        <div class="bar-scrubber" data-volume="" style="">
            <div class="bar-scrubber-icon gcore-skin-main-color" data-volume=""></div>
        </div>
    </div>
    
        </div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-clips=""></div>
    
        </div>
    
    
    
        <div class="media-control-right-panel" data-media-control="">
            
        <button type="button" class="media-control-button media-control-icon gcore-skin-button-color" data-fullscreen="" aria-label="fullscreen">/assets/icons/new/fullscreen-off.svg</button>
    
        <div class="media-control-multicamera" data-multicamera=""></div>
    
        <div class="media-control-vr" data-vr=""></div>
    
        <div class="my-media-control" data-cc="">test</div></div>
    
</div>
<style>:root {}</style>"
`;

exports[`MediaControl > putElement > gear > should put the element in the right panel 1`] = `
"<div class="media-control-background" data-background=""></div>

<div class="media-control-layer gcore-skin-bg-color" data-controls="">
    
    
    
    
    
    
    
    
    

    
    
        <div class="media-control-left-panel" data-media-control="">
            
        <div class="drawer-container" data-volume="">
            <div class="drawer-icon-container" data-volume="">
                <div class="drawer-icon media-control-icon gcore-skin-button-color" data-volume="">/assets/icons/new/volume-max.svg</div>
                <span class="drawer-text" data-volume=""></span>
            </div>
            
    <div class="bar-container" data-volume="">
        <div class="bar-background" data-volume="">
            <div class="bar-fill-1 gcore-skin-main-color" data-volume="" style=""></div>
        </div>
        <div class="bar-scrubber" data-volume="" style="">
            <div class="bar-scrubber-icon gcore-skin-main-color" data-volume=""></div>
        </div>
    </div>
    
        </div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-clips=""></div>
    
        </div>
    
    
    
        <div class="media-control-right-panel" data-media-control="">
            
        <button type="button" class="media-control-button media-control-icon gcore-skin-button-color" data-fullscreen="" aria-label="fullscreen">/assets/icons/new/fullscreen-off.svg</button>
    
        <div class="media-control-multicamera" data-multicamera=""></div>
    
        <div class="media-control-vr" data-vr=""></div>
    
        <div class="my-media-control" data-gear="">test</div></div>
    
</div>
<style>:root {}</style>"
`;

exports[`MediaControl > putElement > pip > should put the element in the right panel 1`] = `
"<div class="media-control-background" data-background=""></div>

<div class="media-control-layer gcore-skin-bg-color" data-controls="">
    
    
    
    
    
    
    
    
    

    
    
        <div class="media-control-left-panel" data-media-control="">
            
        <div class="drawer-container" data-volume="">
            <div class="drawer-icon-container" data-volume="">
                <div class="drawer-icon media-control-icon gcore-skin-button-color" data-volume="">/assets/icons/new/volume-max.svg</div>
                <span class="drawer-text" data-volume=""></span>
            </div>
            
    <div class="bar-container" data-volume="">
        <div class="bar-background" data-volume="">
            <div class="bar-fill-1 gcore-skin-main-color" data-volume="" style=""></div>
        </div>
        <div class="bar-scrubber" data-volume="" style="">
            <div class="bar-scrubber-icon gcore-skin-main-color" data-volume=""></div>
        </div>
    </div>
    
        </div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-clips=""></div>
    
        </div>
    
    
    
        <div class="media-control-right-panel" data-media-control="">
            
        <button type="button" class="media-control-button media-control-icon gcore-skin-button-color" data-fullscreen="" aria-label="fullscreen">/assets/icons/new/fullscreen-off.svg</button>
    
        <div class="media-control-multicamera" data-multicamera=""></div>
    
        <div class="media-control-vr" data-vr=""></div>
    
        <div class="my-media-control" data-pip="">test</div></div>
    
</div>
<style>:root {}</style>"
`;

exports[`MediaControl > rendering timing > once metadata is loaded > should wait a delay before rendering anything 1`] = `
"<div class="media-control-background" data-background=""></div>

<div class="media-control-layer gcore-skin-bg-color" data-controls="">
    
    
    
    
    
    
    
    
    

    
    
        <div class="media-control-left-panel" data-media-control="">
            
        <button type="button" class="media-control-button media-control-icon gcore-skin-button-color paused" data-playpause="" aria-label="playpause">/assets/icons/new/play.svg</button>
    
        <div class="drawer-container" data-volume="">
            <div class="drawer-icon-container" data-volume="">
                <div class="drawer-icon media-control-icon gcore-skin-button-color" data-volume="">/assets/icons/new/volume-max.svg</div>
                <span class="drawer-text" data-volume=""></span>
            </div>
            
    <div class="bar-container" data-volume="">
        <div class="bar-background" data-volume="">
            <div class="bar-fill-1 gcore-skin-main-color" data-volume="" style=""></div>
        </div>
        <div class="bar-scrubber" data-volume="" style="">
            <div class="bar-scrubber-icon gcore-skin-main-color" data-volume=""></div>
        </div>
    </div>
    
        </div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-position="">00:00</div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-duration="">00:00</div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-clips=""></div>
    
        </div>
    
    
        <div class="media-control-center-panel" data-media-control="">
            
        <div class="bar-container" data-seekbar="">
            <div class="bar-background" data-seekbar="">
                <div class="bar-fill-1" data-seekbar=""></div>
                <div class="bar-fill-2 gcore-skin-main-color" data-seekbar="" style=""></div>
                <div class="bar-hover" data-seekbar=""></div>
            </div>
        </div>
        <div class="bar-scrubber" data-seekbar="" style="">
            <div class="bar-scrubber-icon gcore-skin-main-color" data-seekbar=""></div>
        </div>
    
        </div>
    
    
        <div class="media-control-right-panel" data-media-control="">
            
        <button type="button" class="media-control-button media-control-icon gcore-skin-button-color" data-fullscreen="" aria-label="fullscreen">/assets/icons/new/fullscreen-off.svg</button>
    
        <div class="media-control-multicamera" data-multicamera=""></div>
    
        <div class="media-control-vr" data-vr=""></div>
    
        </div>
    
</div>
<style>:root {}</style>"
`;

exports[`MediaControl > updateSettings > dvr > when disabled > should disable DVR controls 1`] = `
"<div class="media-control-background" data-background=""></div>

<div class="media-control-layer gcore-skin-bg-color" data-controls="">
    
    
    
    
    
    
    
    
    

    
    
        <div class="media-control-left-panel" data-media-control="">
            
        <button type="button" class="media-control-button media-control-icon gcore-skin-button-color paused" data-playpause="" aria-label="playpause">/assets/icons/new/play.svg</button>
    
        <div class="drawer-container" data-volume="">
            <div class="drawer-icon-container" data-volume="">
                <div class="drawer-icon media-control-icon gcore-skin-button-color" data-volume="">/assets/icons/new/volume-max.svg</div>
                <span class="drawer-text" data-volume=""></span>
            </div>
            
    <div class="bar-container" data-volume="">
        <div class="bar-background" data-volume="">
            <div class="bar-fill-1 gcore-skin-main-color" data-volume="" style=""></div>
        </div>
        <div class="bar-scrubber" data-volume="" style="">
            <div class="bar-scrubber-icon gcore-skin-main-color" data-volume=""></div>
        </div>
    </div>
    
        </div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-position="">00:00</div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-duration="">00:00</div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-clips=""></div>
    
        </div>
    
    
    
        <div class="media-control-right-panel" data-media-control="">
            
        <button type="button" class="media-control-button media-control-icon gcore-skin-button-color" data-fullscreen="" aria-label="fullscreen">/assets/icons/new/fullscreen-off.svg</button>
    
        <div class="media-control-multicamera" data-multicamera=""></div>
    
        <div class="media-control-vr" data-vr=""></div>
    
        </div>
    
</div>
<style>:root {}</style>"
`;

exports[`MediaControl > updateSettings > dvr > when enabled > should enable DVR controls 1`] = `
"<div class="media-control-background" data-background=""></div>

<div class="media-control-layer gcore-skin-bg-color" data-controls="">
    
    
    
    
    
    
    
    
    

    
    
        <div class="media-control-left-panel" data-media-control="">
            
        <button type="button" class="media-control-button media-control-icon gcore-skin-button-color paused" data-playpause="" aria-label="playpause">/assets/icons/new/play.svg</button>
    
        <div class="drawer-container" data-volume="">
            <div class="drawer-icon-container" data-volume="">
                <div class="drawer-icon media-control-icon gcore-skin-button-color" data-volume="">/assets/icons/new/volume-max.svg</div>
                <span class="drawer-text" data-volume=""></span>
            </div>
            
    <div class="bar-container" data-volume="">
        <div class="bar-background" data-volume="">
            <div class="bar-fill-1 gcore-skin-main-color" data-volume="" style=""></div>
        </div>
        <div class="bar-scrubber" data-volume="" style="">
            <div class="bar-scrubber-icon gcore-skin-main-color" data-volume=""></div>
        </div>
    </div>
    
        </div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-position="">00:00</div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-duration="">00:00</div>
    
        <div class="media-control-indicator gcore-skin-text-color" data-clips=""></div>
    
        <div class="my-dvr-controls" data-dvr="">live</div></div>
    
    
    
        <div class="media-control-right-panel" data-media-control="">
            
        <button type="button" class="media-control-button media-control-icon gcore-skin-button-color" data-fullscreen="" aria-label="fullscreen">/assets/icons/new/fullscreen-off.svg</button>
    
        <div class="media-control-multicamera" data-multicamera=""></div>
    
        <div class="media-control-vr" data-vr=""></div>
    
        </div>
    
</div>
<style>:root {}</style>"
`;
